src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -576,9 +577,31 @@ protected RubyDynamicObject clone(RubyDynamicObject object, boolean freeze,
                 newObjectMetaClass.fields.initCopy(selfMetaClass);
             }
 
-            initializeCloneNode.call(newObject, "initialize_clone", object);
+            // pass :freeze keyword argument to #initialize_clone
+            if (toForceFreezing(freeze) || toForceUnfreezing(freeze)) {
+                final String string = "freeze";
+                final LeafRope rope = RopeOperations.encodeAscii(string, USASCIIEncoding.INSTANCE);
+                final RubySymbol key = new RubySymbol(string, rope, Encodings.US_ASCII);
+                final boolean value = toForceFreezing(freeze);
+
+                final Object[] newStore = PackedHashStoreLibrary.createStore();
+                final int hashed = hashNode.execute(key);
+                PackedHashStoreLibrary.setHashedKeyValue(newStore, 0, hashed, key, value);
+
+                final RubyHash hash = new RubyHash(
+                        coreLibrary().hashClass,
+                        getLanguage().hashShape,
+                        getContext(),
+                        newStore,
+                        1);
```

Comment:
I am not familiar with the Rope concept and how to build it properly for building a Symbol. So any example or clarification will be very helpful 🙇 .

```ruby
PackedHashStoreLibrary.setHashedKeyValue(newStore, 0, hashed, key, value);
```
I assume here that `0` is an index of the first element in a 0-based array.

```ruby
new RubyHash(
    coreLibrary().hashClass,
    getLanguage().hashShape,
    getContext(),
    newStore,
    1);
```

Here I suppose `1` means how many elements are in the Hash

Also, I suppose I work with Symbols in the wrong way - a new symbol should be probably registered in some repository but not just created. Please advise me how to do it properly 🙇.

Reply:
Yes, that's all correct and looks good.